pkg/controller/elasticsearch/driver/upscale.go

```diff
@@ -66,33 +69,70 @@ func HandleUpscaleAndSpecChanges(
 	if err != nil {
 		return results, fmt.Errorf("adjust resources: %w", err)
 	}
-	// reconcile all resources
-	for _, res := range adjusted {
-		res := res
-		if err := settings.ReconcileConfig(ctx.parentCtx, ctx.k8sClient, ctx.es, res.StatefulSet.Name, res.Config, ctx.meta); err != nil {
-			return results, fmt.Errorf("reconcile config: %w", err)
-		}
-		if _, err := common.ReconcileService(ctx.parentCtx, ctx.k8sClient, &res.HeadlessService, &ctx.es); err != nil {
-			return results, fmt.Errorf("reconcile service: %w", err)
-		}
-		if actualSset, exists := actualStatefulSets.GetByName(res.StatefulSet.Name); exists {
-			recreateSset, err := handleVolumeExpansion(ctx.parentCtx, ctx.k8sClient, ctx.es, res.StatefulSet, actualSset, ctx.validateStorageClass)
-			if err != nil {
-				return results, fmt.Errorf("handle volume expansion: %w", err)
-			}
-			if recreateSset {
-				// The StatefulSet is scheduled for recreation: let's requeue before attempting any further spec change.
-				results.Requeue = true
-				continue
-			}
-		}
-		reconciled, err := es_sset.ReconcileStatefulSet(ctx.parentCtx, ctx.k8sClient, ctx.es, res.StatefulSet, ctx.expectations)
+
+	// Check if this is a version upgrade
+	isVersionUpgrade, err := isVersionUpgrade(ctx.es)
+	if err != nil {
+		return results, fmt.Errorf("while checking for version upgrade: %w", err)
+	}
+
+	// If this is not a version upgrade, process all resources normally and return
+	if !isVersionUpgrade {
+		actualStatefulSets, requeue, err := reconcileResources(ctx, actualStatefulSets, adjusted)
 		if err != nil {
-			return results, fmt.Errorf("reconcile StatefulSet: %w", err)
+			return results, fmt.Errorf("while reconciling resources: %w", err)
 		}
-		// update actual with the reconciled ones for next steps to work with up-to-date information
-		actualStatefulSets = actualStatefulSets.WithStatefulSet(reconciled)
+		results.Requeue = requeue
+		results.ActualStatefulSets = actualStatefulSets
+		return results, nil
 	}
+
+	// Version upgrade: separate master and non-master StatefulSets
+	var masterResources, nonMasterResources []nodespec.Resources
+	for _, res := range adjusted {
+		if label.IsMasterNodeSet(res.StatefulSet) {
+			masterResources = append(masterResources, res)
+		} else {
+			nonMasterResources = append(nonMasterResources, res)
+		}
+	}
+
+	// First, reconcile all non-master resources
+	actualStatefulSets, requeue, err := reconcileResources(ctx, actualStatefulSets, nonMasterResources)
+	if err != nil {
+		return results, fmt.Errorf("while reconciling non-master resources: %w", err)
+	}
+	if requeue {
+		results.Requeue = true
+		results.ActualStatefulSets = actualStatefulSets
+		return results, nil
```

Comment:
Not sure how much it is a problem, but I think this prevents the masters from being scaled until all the other nodes have been upgraded (this can be done today). I don't know if this could be a problem on large clusters, if a user wants to give more capacity to the ES control plane during an upgrade.